### DISCUSSION TITLE
java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found

### DISCUSSION THREAD

**Discussion by euronymous** [maruTA-bis5/mattermost4j#482] [Q&A]
Hi,

I don't know java stuff. I'm guessing that I'm doing something wrong...
I'm still getting:
`java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found`
```
$ java -version
openjdk version "11.0.15" 2022-04-19
```

I'm building a fat jar of course.
Iterated through many different SO topics/solutions. Added many different deps:
```
        "jakarta.activation"          % "jakarta.activation-api"    % "2.0.0",
        "jakarta.xml.bind"            % "jakarta.xml.bind-api"      % "4.0.0",
        "jakarta.ws.rs"               % "jakarta.ws.rs-api"         % "3.1.0",
        "org.glassfish.jersey.core"   % "jersey-client"             % "3.0.4",
        "org.glassfish.jersey.inject" % "jersey-hk2"                % "3.0.4",
        "org.glassfish.jersey.media"  % "jersey-media-json-binding" % "3.0.4",
        "com.fasterxml.jackson.core"  % "jackson-databind"          % "2.13.3",
        "org.glassfish.jersey"        % "jersey-bom"                % "3.0.4",
        "org.jboss.resteasy"          % "resteasy-client"           % "3.0.2.Final"
```

Then I've added the "activation api" and "xml bindings" (whatever that is) to the cp.
Also haven't helped...
 
```
$ java -cp ~/.cache/coursier/v1/https/repo1.maven.org/maven2/jakarta/xml/bind/jakarta.xml.bind-api/4.0.0/jakarta.xml.bind-api-4.0.0.jar:~/.cache/coursier/v1/https/repo1.maven.org/maven2/jakarta/ws/rs/jakarta.ws.rs-api/3.0.0/jakarta.ws.rs-api-3.0.0.jar:~/.cache/coursier/v1/https/repo1.maven.org/maven2/jakarta/activation/jakarta.activation-api/2.1.0/jakarta.activation-api-2.1.0.jar:~/mattermost-notifier/target/scala-2.13/mattermost-notifier-assembly-0.1.0-SNAPSHOT.jar pl.softwerk.mattermost_notifier.Main 
Exception in thread "main" java.lang.RuntimeException: java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found
	at jakarta.ws.rs.client.ClientBuilder.newBuilder(ClientBuilder.java:75)
	at net.bis5.mattermost.client4.MattermostClient.buildClient(MattermostClient.java:200)
	at net.bis5.mattermost.client4.MattermostClient.<init>(MattermostClient.java:272)
	at net.bis5.mattermost.client4.MattermostClient$MattermostClientBuilder.build(MattermostClient.java:245)
	at pl.softwerk.mattermost_idle.Main$.run(Main.scala:17)
	at cats.effect.IOApp.main(IOApp.scala:268)
	at cats.effect.IOApp.main$(IOApp.scala:203)
	at pl.softwerk.mattermost_notifier.Main$.main(Main.scala:13)
	at pl.softwerk.mattermost_notifier.Main.main(Main.scala)
Caused by: java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found
	at jakarta.ws.rs.client.FactoryFinder.find(FactoryFinder.java:163)
	at jakarta.ws.rs.client.ClientBuilder.newBuilder(ClientBuilder.java:62)
	... 8 more
```

**Comment by maruTA-bis5** [comment]
1. Your `"jakarta.ws.rs" % "jakarta.ws.rs-api" % "3.1.0"` is quite new. The `"jersey-client" % "3.0.4"` implements ws.rs-api `3.0.0`.
1. Mattermost4j doesn't support RESTEasy currently. It will be available by PR #450 after the release RESTEasy 6.1.0.Final.
So you don't need `"org.jboss.resteasy"          % "resteasy-client"           % "3.0.2.Final"`.

If these dependency changes haven't helped, I don't know about scala and its build system, but I think that your fat jar (or your classpath) does not contain implementation of `jakarta.ws.rs.client.ClientBuilder`.

**Comment by euronymous** [comment]
Hi, thanks for your answer. Sorry, I thought about creating a discussion instead of issue... nvm :)

1. I've tried with the `rs-api` @ `3.0.0` version in the first place :(
2. That dep was suggested to a similar issue. Didn't helped ofc.

Also checked the presence of ClientBuilder:
```
$ jar tf mattermost-notifier-assembly-0.1.0-SNAPSHOT.jar | grep ClientBuilder
jakarta/ws/rs/client/ClientBuilder.class
javax/ws/rs/client/ClientBuilder.class
net/bis5/mattermost/client4/MattermostClient$MattermostClientBuilder.class
org/glassfish/jersey/client/JerseyClientBuilder$1.class
org/glassfish/jersey/client/JerseyClientBuilder.class
org/glassfish/jersey/client/spi/ClientBuilderListener.class
org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder$1.class
org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder$HostnameVerificationPolicy.class
org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder$VerifierWrapper.class
org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder.class
```
But... important word: "implementation" :)
Is it okay, that there is only an abstract class? Shouldn't there be any implementation of it?

I also thought that maybe adding some implementation of that ClientBuilder would solve the problem.
But... he is complaining strictly about the `java.lang.ClassNotFoundException: Provider for jakarta.ws.rs.client.ClientBuilder cannot be found`.

**Reply by maruTA-bis5** [reply]
The `Provider for jakarta.ws.rs.client.ClientBuilder` is loaded by the ServiceProvider. 
You need `META-INF/services/jakarta.ws.rs.client.ClientBuilder`(It's included in `jersey-client`'s jar).

`jakarta/ws/rs/client/ClientBuilder.class` is the API provided by `rs-api`. 
The specification is as known as "Jakarta RESTful Web Services" or "JAX-RS". It is a part of Jakarta EE.

`org/glassfish/jersey/client/JerseyClientBuilder.class` is the implementation of `ClientBuilder` provided by `jersey-client`. It is a part of [Jersey](https://github.com/eclipse-ee4j/jersey).
`org/jboss/resteasy/client/jaxrs/ResteasyClientBuilder.class` is also the implementation of `ClientBuilder` provided by `resteasy-client`. It is a part of [RESTEasy](https://resteasy.dev/).
*) Jersey and RESTEasy: Both provide "implementation" of JAX-RS. Note that RESTEasy is not supported by Mattermost4J currently.
